File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/tracker/InstallTracker.java
package com.snowplowanalytics.snowplow.internal.tracker;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.preference.PreferenceManager;

import androidx.annotation.NonNull;
import androidx.annotation.RestrictTo;

import com.snowplowanalytics.snowplow.event.SelfDescribing;
import com.snowplowanalytics.snowplow.internal.constants.TrackerConstants;
import com.snowplowanalytics.snowplow.internal.utils.NotificationCenter;
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson;

import java.util.Calendar;
import java.util.HashMap;
import java.util.Map;

import static com.snowplowanalytics.snowplow.internal.constants.TrackerConstants.INSTALLED_BEFORE;
import static com.snowplowanalytics.snowplow.internal.constants.TrackerConstants.INSTALL_TIMESTAMP;

/**
 * Class used to keep track of install state of app.
 * If a file does not exist, the tracker will send an `application_install` event.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
public class InstallTracker {
    private static String TAG = InstallTracker.class.getSimpleName();

    private Boolean isNewInstall;
    private SharedPreferences sharedPreferences;

    private static InstallTracker sharedInstance;

    @NonNull
    public synchronized static InstallTracker getInstance(@NonNull Context context) {
        if (sharedInstance == null) {
            sharedInstance = new InstallTracker(context);
        }
        return sharedInstance;
    }

    private InstallTracker(@NonNull Context context) {
        new SharedPreferencesTask().execute(context);
    }

    private class SharedPreferencesTask extends AsyncTask<Context, Void, Boolean> {
        @Override
        protected Boolean doInBackground(Context... contexts) {
            sharedPreferences = PreferenceManager.getDefaultSharedPreferences(contexts[0]);
            if (sharedPreferences.getString(INSTALLED_BEFORE, null) == null) {
                // mark the install if there's no value
                SharedPreferences.Editor editor = sharedPreferences.edit();
                editor.putString(INSTALLED_BEFORE, "YES");
                editor.putLong(INSTALL_TIMESTAMP, Calendar.getInstance().getTimeInMillis());
                editor.apply();
                // since the value was missing in sharedPreferences, we're assuming this is a new install
                isNewInstall = true;
            } else {
                // if there's an INSTALLED_BEFORE record in sharedPreferences - someone has been there!
                isNewInstall = false;
            }
            return isNewInstall;
        }

        @Override
        protected void onPostExecute(Boolean isNewInstall) {
            long installTimestamp = sharedPreferences.getLong(INSTALL_TIMESTAMP, 0);
            // We send the installEvent if it's a new installed app but in case the tracker hasn't been able
            // to send the event before we can retry checking if INSTALL_TIMESTAMP was already removed.
            if (!isNewInstall && installTimestamp <= 0) {
                return;
            }
            sendInstallEvent(installTimestamp);
            // clear install timestamp
            SharedPreferences.Editor editor = sharedPreferences.edit();
            editor.remove(INSTALL_TIMESTAMP);
            editor.commit();
        }
    }

    private void sendInstallEvent(long installTimestamp) {
        SelfDescribing event = new SelfDescribing(new SelfDescribingJson(TrackerConstants.SCHEMA_APPLICATION_INSTALL));
        if (installTimestamp > 0) {
            event.trueTimestamp(installTimestamp);
        }
        Map<String, Object> notificationData = new HashMap<String, Object>();
        notificationData.put("event", event);
        NotificationCenter.postNotification("SnowplowInstallTracking", notificationData);
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/tracker/InstallTracker.kt
/*
 * Copyright (c) 2015-2023 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.core.tracker

import android.content.Context
import android.content.SharedPreferences
import android.os.AsyncTask
import androidx.preference.PreferenceManager
import androidx.annotation.RestrictTo
import com.snowplowanalytics.core.constants.TrackerConstants
import com.snowplowanalytics.core.utils.NotificationCenter.postNotification
import com.snowplowanalytics.snowplow.event.SelfDescribing
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson
import java.util.*

/**
 * Class used to keep track of install state of app.
 * If a file does not exist, the tracker will send an `application_install` event.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
class InstallTracker private constructor(context: Context) {
    private var isNewInstall: Boolean? = null
    private var sharedPreferences: SharedPreferences? = null

    init {
        SharedPreferencesTask().execute(context)
    }

    private inner class SharedPreferencesTask : AsyncTask<Context?, Void?, Boolean>() {
        
        override fun doInBackground(vararg params: Context?): Boolean? {
            sharedPreferences = params[0]?.let { PreferenceManager.getDefaultSharedPreferences(it) }
            isNewInstall = if (sharedPreferences?.getString(TrackerConstants.INSTALLED_BEFORE, null) == null) {
                // mark the install if there's no value
                val editor = sharedPreferences?.edit()
                editor?.putString(TrackerConstants.INSTALLED_BEFORE, "YES")
                editor?.putLong(TrackerConstants.INSTALL_TIMESTAMP, Calendar.getInstance().timeInMillis)
                editor?.apply()
                // since the value was missing in sharedPreferences, we're assuming this is a new install
                true
            } else {
                // if there's an INSTALLED_BEFORE record in sharedPreferences - someone has been there!
                false
            }
            return isNewInstall
        }

        override fun onPostExecute(isNewInstall: Boolean) {
            val installTimestamp = sharedPreferences?.getLong(TrackerConstants.INSTALL_TIMESTAMP, 0)
            // We send the installEvent if it's a new installed app but in case the tracker hasn't been able
            // to send the event before we can retry checking if INSTALL_TIMESTAMP was already removed.
            installTimestamp?.let {
                if (!isNewInstall && installTimestamp <= 0) {
                    return
                }
                sendInstallEvent(it) 
            }
            // clear install timestamp
            val editor = sharedPreferences?.edit()
            editor?.remove(TrackerConstants.INSTALL_TIMESTAMP)
            editor?.commit()
        }
    }

    private fun sendInstallEvent(installTimestamp: Long) {
        val event = SelfDescribing(SelfDescribingJson(TrackerConstants.SCHEMA_APPLICATION_INSTALL))
        if (installTimestamp > 0) {
            event.trueTimestamp(installTimestamp)
        }
        val notificationData: MutableMap<String, Any> = HashMap()
        notificationData["event"] = event
        postNotification("SnowplowInstallTracking", notificationData)
    }

    companion object {
        private val TAG = InstallTracker::class.java.simpleName
        private var sharedInstance: InstallTracker? = null
        
        @JvmStatic
        @Synchronized
        fun getInstance(context: Context): InstallTracker {
            if (sharedInstance == null) {
                sharedInstance = InstallTracker(context)
            }
            return sharedInstance!!
        }
    }
}
